torch/distributed/nn/functional.py

```diff
@@ -262,31 +264,38 @@ class _Reduce_Scatter(Function):
     @staticmethod
     def forward(ctx, op, group, tensor, *input_tensor_list):
         ctx.group = group
+        input_tensor_list = tuple(t.contiguous() for t in input_tensor_list)
         dist.reduce_scatter(tensor, list(input_tensor_list), op=op, group=group)
         return tensor
 
     @staticmethod
     def backward(ctx, grad_output):
-        return (None, None, None) + _AllGather.apply(
-            ctx.group, grad_output.contiguous()
-        )
+        return (None, None, None) + _AllGather.apply(ctx.group, grad_output)
 
 
 class _AllGather(Function):
     @staticmethod
     def forward(ctx, group, tensor):
         ctx.group = group
         out_tensor_list = [
-            torch.empty_like(tensor) for i in range(dist.get_world_size(group=group))
+            torch.empty_like(tensor) for _ in range(dist.get_world_size(group=group))
         ]
-        dist.all_gather(out_tensor_list, tensor, group=group)
+
+        dist.all_gather(out_tensor_list, tensor.contiguous(), group=group)
```

Comment:
Hi @SherlockNoMad!

I believe this does not resolve the issue, because `out_tensor_list` is not continues since it uses `empty_like`

The correct solution would be

```py
        tensor = tensor.contiguous()
        out_tensor_list = [
            torch.empty_like(tensor) for _ in range(dist.get_world_size(group=group))
        ]
        dist.all_gather(out_tensor_list, tensor, group=group)
```

The same comment applies to the other classes